src/bytecode.rs

```diff
@@ -1,69 +1,301 @@
-    Not { dest: Register, src: Register },
-    Equal { dest: Register, lhs: Register, rhs: Register },
-    NotEqual { dest: Register, lhs: Register, rhs: Register },
-    Greater { dest: Register, lhs: Register, rhs: Register },
-    GreaterEqual { dest: Register, lhs: Register, rhs: Register },
-    Less { dest: Register, lhs: Register, rhs: Register },
-    LessEqual { dest: Register, lhs: Register, rhs: Register },
-    Jump { offset: JumpOffset },
-    JumpIfFalse { condition: Register, offset: JumpOffset },
-    Call { dest: Register, function: FunctionId, args: Vec<Register> },
-    Return { src: Register },
-    GetGlobal { dest: Register, name_idx: ConstantIndex },
-    SetGlobal { name_idx: ConstantIndex, src: Register },
-    GetLocal { dest: Register, slot: u8 }, // Assuming locals are slot-based
-    SetLocal { slot: u8, src: Register },
-    MakeObject { dest: Register, type_idx: ConstantIndex },
-    GetField { dest: Register, object: Register, field_idx: ConstantIndex },
-    SetField { object: Register, field_idx: ConstantIndex, src: Register },
-    Halt,
+pub enum Operand {
+    Register(usize),                   // A register (r0, r1, etc.)
+    StackAddr(i32),                    // Stack pointer offset [sp+n]
+    Immediate(Value),                  // An immediate value (constant)
+    Label(String),                     // A jump label
+}
+
+/// Resolved instruction with numeric indices instead of strings
+///
+/// This is an optimized representation of instructions used during execution.
+/// String identifiers (like function names and labels) are replaced with numeric indices,
+/// allowing for faster execution without string lookups.
+///
+/// This representation is generated from the symbolic `Instruction` enum during function
+/// registration and is used by the VM's execution engine.
+#[derive(Debug, Clone, Copy)]
+pub enum ResolvedInstruction {
+    // Load value from stack to register
+    LD(usize, i32),                      // LD r1, [sp+0]
+    // Load immediate value to register
+    LDI(usize, usize),                   // LDI r1, const_idx (register, constant index)
+    // Move value from one register to another
+    MOV(usize, usize),                   // MOV r1, r2
+    // Arithmetic operations
+    ADD(usize, usize, usize),            // ADD r1, r2, r3
+    SUB(usize, usize, usize),            // SUB r1, r2, r3
+    MUL(usize, usize, usize),            // MUL r1, r2, r3
+    DIV(usize, usize, usize),            // DIV r1, r2, r3
+    MOD(usize, usize, usize),            // MOD r1, r2, r3
+    // Bitwise operations
+    AND(usize, usize, usize),            // AND r1, r2, r3
+    OR(usize, usize, usize),             // OR r1, r2, r3
+    XOR(usize, usize, usize),            // XOR r1, r2, r3
+    NOT(usize, usize),                   // NOT r1, r2
+    SHL(usize, usize, usize),            // SHL r1, r2, r3
+    SHR(usize, usize, usize),            // SHR r1, r2, r3
+    // Control flow
```

Comment:
This might end up being removed. Can't do jumps in MPC VMs without leaking which path was taken.

Reply:
So, in MPC, you can't do conditionals over secret-shared values as you can in regular programming languages anyway. You always have to execute both branches. Otherwise, you leak which branch executed. Either you reveal the secret shared value and then do a conditional over that or do a MUX. 
As discussed on TG, there are generalizations of MUX that we can use as a kind of redirection sugar for these JUMP instructions. What might make sense is to implement that and then once we get more feedback (our current customers and upcoming beta testers), we can decide if this sugar makes sense to keep or not.